assets/src/edit-story/output/utils/getFontDeclarations.js

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import getGoogleFontURL from './getGoogleFontURL';
+
+/**
+ * Returns a list of font declarations across all pages in a story.
+ *
+ * @param {import('../../types').Page[]} pages List of pages.
+ *
+ * @return {string[]} Font declarations.
+ */
+const getFontDeclarations = (pages) => {
+  const map = new Map();
+
+  for (const { elements } of pages) {
+    const textElements = elements.filter(({ type }) => type === 'text');
+
+    // Prepare font objects for later use.
+    for (const { font, fontStyle, fontWeight } of textElements) {
+      const { service, family } = font;
+      if (!service || service === 'system') {
+        continue;
+      }
+
+      const serviceMap = map.get(service) || new Map();
+      map.set(service, serviceMap);
+
+      // Example: [1, 700] for bold + italic.
+      const variant = [Number(fontStyle === 'italic'), fontWeight || 400];
```

Comment:
But what if the font doesn't actually have a special "italic" variant or a specified weight? E.g. I just tried https://fonts.googleapis.com/css2?display=swap&family=Architects+Daughter:ital,wght@1,300;1,900 and it doesn't work.

Reply:
We‘d need to make sure the user cannot set invalid combinations in the design panel. Related: #876

Reply:
I don't think that's how this works. It's ok to specify `font-style: italic` or `font-weight: 900` on almost any text. If there's no a matching `@font-face {}`, then the browser does it's own math to do bold/italic/bolditalic. And if the matching `@font-faces` are available, a browser would do some optimal selections. For instance:

* If a `@font-face {font-weight: 700}` is available and `p {font-weight: 700}` is displayed - this font face will be used - exact match.
* If a `@font-face {font-weight: 700}` is available and `p {font-weight: 900}` is displayed - the 700 font face will still be used (nearest `<= 900`). A browser may "add weight" to this font programmatically, but I don't think any do today.
* If a `@font-face {font-weight: 900}` is available and `p {font-weight: 700}` is displayed - the 900 font face will still be used (nearest `>= 700`) even though it's a bigger weight.

Thus I think our selection rules should be somewhat similar.

Reply:
So when the user makes a certain selection, we can use the `FontProvider` to see if the font supports that combination and if so, update the `font` object accordingly. It's not something we can handle during output.

Reply:
The editor should do the best with what it has. But the output should still ask for the most appropriate font-variant combinations based on what the content does. Again, I don't think it's ever wrong to have a `font-weight: 900` - the browser would do a good job even if we don't have an optimal font. The same with italics. I expect most of fonts do not have an italic variant, but it's a valid formatting nonetheless. But I don't believe we can ask Google Fonts to give us italic in that case - I'm pretty sure it will simply fail.

Reply:
So far so clear, but during output we don't know whether a `@font-face {font-weight: 900}` is available for a font. Is that why you're suggesting including `weights` in the `font` object?

Also, FWIW, I don't think we allow setting `font-weight: 900` in the editor unless a font supports it.

Reply:
Yep. That's why I think we should have `weights` in the font object, and `styles` too. Alternatively we can load the whole database of fonts before we start outputting. But I'd much rather prefer if we had `weights` and `styles` in the `font{}`.

> Also, FWIW, I don't think we allow setting font-weight: 900 in the editor unless a font supports it.

That's true. But we do allow setting `700` (normal bold), and italic and these could be absent in the variants too. Take a look at the [Architect's Daughter](https://fonts.google.com/specimen/Architects+Daughter?query=architect&sidebar.open&selection.family=Architects+Daughter) font.  It has only one variant - regular 400.  But you can apply bold(700) and/or italic to it as seen [here](https://output.jsbin.com/qelofus/quiet). However, if I [try to request 700 or italic via API](https://fonts.googleapis.com/css2?family=Architects+Daughter:ital@1&display=block) - I get 400.